### PR TITLE
Extraction macros for ProVerif

### LINKED ISSUE #560: [ProVerif] Annotation macros for constructors, handwritten models

### PR DESCRIPTION
**Pull request by jschneider-bensch**
This PR addresses #560, by introducing macros to annotate functions as either needing handwritten models or that they should be extracted as constructors of their return types.

~It's a draft at the moment, since it builds on changes in #550.~
Fixes #560

### CODE REVIEW

File: hax-lib-macros/src/lib.rs [hunk 1]
```diff
@@ -520,3 +520,21 @@ pub fn protocol_messages(_attr: pm::TokenStream, item: pm::TokenStream) -> pm::T
...
+/// A marker indicating a `fn` as a ProVerif process initialization.
+#[proc_macro_error]
+#[proc_macro_attribute]
+pub fn pv_constructor(_attr: pm::TokenStream, item: pm::TokenStream) -> pm::TokenStream {
+    let item: ItemFn = parse_macro_input!(item);
+    let attr = AttrPayload::PVConstructor;
+    quote! {#attr #item}.into()
+}
+
+/// A marker indicating a `fn` as a ProVerif process initialization.
```
Comment (W95Psp) [inline hax-lib-macros/src/lib.rs:533]: Are the doc comments up to date here?

File: hax-lib-macros/types/src/lib.rs [hunk 2]
```diff
@@ -54,6 +54,8 @@ pub enum AssociationRole {
     ProcessWrite,
     ProcessInit,
     ProtocolMessages,
+    PVConstructor,
+    PVHandwritten,
```
Comment (W95Psp) [inline hax-lib-macros/types/src/lib.rs:58]: Do you actually need constructors both in AssociationRole and in AttrPayload?
`AssociationRole` defines the nature of the relation of two linked items, e.g. a function `f` is linked to `f_precondition` (via a UUID) and the nature of their relation is "Requires" (I need to write a comment for that :sweat_smile:)

File: hax-lib-macros/src/lib.rs (hunk 1 above)
Comment (jschneider-bensch) [reply hax-lib-macros/src/lib.rs:533]: Good catch, they're not!

File: hax-lib-macros/types/src/lib.rs (hunk 2 above)
Comment (jschneider-bensch) [reply hax-lib-macros/types/src/lib.rs:58]: I see, yes, then `AssociationRole` is not necessary at all, since there are no associated items.

**Review by W95Psp — APPROVED** [review]
Lgtm :)

**Pull request merged**